metricbeat/module/mssql/db/db.go

```diff
@@ -0,0 +1,176 @@
+package db
+
+import (
+	"database/sql"
+	"fmt"
+	_ "github.com/denisenkom/go-mssqldb"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/module/mssql"
+	"github.com/pkg/errors"
+	"sync"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "db", New, mb.DefaultMetricSet())
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*mssql.MetricSet
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The mssql io metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	//TODO Db must be gracefully closed
+
+	metricSet, err := mssql.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating mssql metricset")
+	}
+
+	ms := &MetricSet{MetricSet: metricSet}
+
+	return ms, nil
+}
+
+func newConcurrentQueries(db *sql.DB, qs []string) *concurrentResults {
+	qr := &concurrentResults{
+		queries: qs,
+		db:      db,
+	}
+
+	qr.Add(len(qs))
+
+	return qr
+}
+
+type concurrentResults struct {
+	queries []string
+	maprs   []common.MapStr
+	error
+	sync.Mutex
+	sync.WaitGroup
+	db *sql.DB
+}
+
+func (c *concurrentResults) doQueries() {
+	for _, q := range c.queries {
+		go c.doEventsFetcherQuery(fmt.Sprintf("SELECT * FROM sys.%s;", q))
+	}
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch() ([]common.MapStr, error) {
```

Comment:
There are multiple interfaces a metricset can implement depending on the signature of its `Fetch` method, this one returning the event is the "old" one, we currently prefer the [`ReportingMetricSetV2` interface](https://godoc.org/github.com/elastic/beats/metricbeat/mb#ReportingMetricSetV2), where the Fetch method receives a `ReporterV2` object that can be used to send any number of events and errors.

Take a look to its usage in other modules, it'd be nice to use it on new modules like this one.